torch/csrc/jit/serialization/export_bytecode.cpp

```diff
@@ -1,11 +1,331 @@
+
 namespace torch {
 namespace jit {
 
+std::vector<Method> gatherMethodsToExport(
+    const Module& module,
+    bool incl_setstate,
+    bool incl_interface_call) {
+  std::vector<Method> methods_to_export;
+  std::deque<Method> stack;
+
+  for (const auto& method : module.get_methods()) {
+    stack.emplace_back(method);
+  }
+
+  if (incl_setstate) {
+    // Use DFS on IValue's to traverse dependencies of module._ivalue and
+    // add all setstate/getstates to initial stack.
+    std::vector<ObjectPtr> ivalue_stack;
+    ivalue_stack.push_back(module._ivalue());
+    while (!ivalue_stack.empty()) {
+      ObjectPtr cur = ivalue_stack.back();
+      ivalue_stack.pop_back();
+      auto type = cur->type();
+      Function* setstate = type->findMethod("__setstate__");
+      Function* getstate = type->findMethod("__getstate__");
+      if (getstate && setstate) {
+        if (setstate->isGraphFunction()) {
+          stack.emplace_back(cur, setstate);
+        }
+        if (getstate->isGraphFunction()) {
+          stack.emplace_back(cur, getstate);
+        }
+      } else {
+        for (size_t i = 0, n = type->numAttributes(); i < n; ++i) {
+          IValue field = cur->getSlot(i);
+          if (field.isObject()) {
+            ivalue_stack.push_back(field.toObject());
+          }
+        }
+      }
+    }
+  }
+
+  // After getting the initial set of functions to export;
+  // recursively search for those that these function calls.
+  // BFS to cover ALL methods of submodules that are used
+  std::set<std::pair<std::string, Function*>> visited;
+  while (!stack.empty()) {
+    Method cur = stack.front();
+    stack.pop_front();
+    auto tup = std::make_pair(
+        cur.owner()._ivalue()->type()->name()->qualifiedName(),
+        &cur.function());
+    if (visited.find(tup) != visited.end()) {
+      continue;
+    }
+    visited.insert(tup);
+    methods_to_export.emplace_back(cur);
+
+    if (incl_interface_call) {
+      const auto& f = toGraphFunction(cur.function());
+      auto graph = f.graph()->copyUnique();
+      Inline(*graph);
+      std::unordered_set<c10::string_view> called_method_names;
+      auto nodes = findAllNodes(*graph, c10::prim::CallMethod, true);
+      for (Node* node : nodes) {
+        if (auto iface = node->input(0)->type()->castRaw<InterfaceType>()) {
+          const FunctionSchema* schema = iface->getMethod(node->s(attr::name));
+          called_method_names.insert(schema->name());
+        }
+      }
+
+      for (const auto& submodule : cur.owner().modules()) {
+        for (const auto& m : submodule.get_methods()) {
+          if (called_method_names.find(m.function().qualname().name()) !=
+              called_method_names.end()) {
+            stack.emplace_back(m);
+          }
+        }
+      }
+    }
+  }
+  return methods_to_export;
+}
+
+std::unique_ptr<mobile::Code> compileGraphToMobileCode(
+    const std::string& name,
+    const std::shared_ptr<Graph>& graph,
```

Comment:
It'd great if we also update `MobileCode` to only take `const Graph&`, but that can be in a separate diff.